src/sentry/services/hybrid_cloud/user/impl.py

```diff
@@ -182,12 +186,25 @@ def get_or_create_user_by_email(self, *, email: str) -> RpcUser:
                     email=email,
                     name=email,
                 )
+                user_signup.send_robust(
+                    sender=self, user=user, source="api", referrer=referrer or "unknown"
+                )
             else:
                 # Users are not supposed to have the same email but right now our auth pipeline let this happen
-                # So let's not break the user experience
+                # So let's not break the user experience. Instead return the user with auth identity of ident or
+                # the first user if ident is None
+                user = user_query[0]
                 if user_query.count() > 1:
                     logger.warning("Email has multiple users", extra={"email": email})
-                user = user_query[0]
+                    if ident:
+                        for candid_user in user_query:
+                            identity_query = AuthIdentity.objects.filter(
```

Comment:
@sentaur-athena This might not work.

We call `auth_service.create_auth_identity()` after `get_or_create_new_user()`:

- `provision_with_new_org()`: https://github.com/getsentry/getsentry/blob/db817ebe096f2f807a52203079760c2aba31c550/getsentry/web/channel_provisioning/process.py#L281-L286
- `provision_for_existing_org()`: https://github.com/getsentry/getsentry/blob/db817ebe096f2f807a52203079760c2aba31c550/getsentry/web/channel_provisioning/process.py#L304-L310

Reply:
@dashed I'm aware of that. The problem I'm facing is for the people who have been in this flow before and already have auth_identity created for them. For example:
1. Athena has 2 users, Athena1 and Athena2
2. She provisions a new org and in that process we create Athena1 an auth identity related to that org
3. Later she provisions a project get_or_create_new_user() returns Athena2 so then auth_service.create_auth_identity() tries to create auth_identity because Athena2 doesn't have one. That would throw an error because Athena1 is using the same ident.

In this fix I want to make sure that if a user already has an ident return that one from get_or_create_new_user() otherwise it doesn't matter.

Reply:
@sentaur-athena that makes sense. The case you outlined might be good to add to https://github.com/getsentry/getsentry/blob/master/tests/getsentry/web/channel_provisioning/test_process.py in case you haven't done so. 👍